File: 53.最大子序和.php
<?php
/*
 * @lc app=leetcode.cn id=53 lang=php
 *
 * [53] 最大子序和
 */
class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer
     */
    function maxSubArray($nums) {
        if (empty($nums)) {
            return 0;
        } else if (1 == count($nums)) {
            return $nums[0];
        }
        $sum = -PHP_INT_MAX;
        $tmpSum = 0;
        $len = count($nums);
        
        for ($i=0; $i < $len; $i++) {
            $tmpSum += $nums[$i];
            if ($sum < $tmpSum) {
                $sum = $tmpSum;
            }
            if ($tmpSum < 0) {
                $tmpSum = 0;
            }
        }

        return $sum;
    }
}

$obj = new Solution();
$arr = array(-2,-1);
$res = $obj->maxSubArray($arr);
var_dump($res);

File: 53.最大子序和.go
/*
 * @lc app=leetcode.cn id=53 lang=golang
 *
 * [53] 最大子序和
 */

// @lc code=start
func maxSubArray(nums []int) int {
    if len(nums) == 0 {
		return 0
	}

	maxNum := nums[0]
	curNum := nums[0]
	len := len(nums)

	for i := 1; i < len; i++ {
		if curNum + nums[i] >= nums[i] {
			curNum = curNum + nums[i]
		} else {
			curNum = nums[i]
		}

		if curNum > maxNum {
			maxNum = curNum
		}
	}

	return maxNum
}
// @lc code=end

